From fa8c26dc098278645163d3d1648005db5cfedd54 Mon Sep 17 00:00:00 2001
From: Gustavo Duarte <gduarte@redhat.com>
Date: Tue, 4 Nov 2014 22:23:04 -0500
Subject: [powerpc] kexec: adjust crashkernel reservation for 2GB-4GB systems

Message-id: <1415139784-28504-1-git-send-email-gduarte@redhat.com>
Patchwork-id: 99397
O-Subject: [RHEL7.1 PATCH BZ 1074924] ppc64: Change crashkernel auto memory reservation value for systems with 2GB-4GB memory
Bugzilla: 1074924
RH-Acked-by: Chao WANG <chaowang@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Baoquan He <bhe@redhat.com>

BZ: 1074924
Upstream Status: RHEL only
Build Info: https://brewweb.devel.redhat.com/taskinfo?taskID=8166150
ppc64le: https://brewweb.devel.redhat.com/taskinfo?taskID=8166151
Tested: IBM verified it fixes the issue using a RHEL7.1 pre-alpha install.

When crashkernel auto memory reservation is used on a system with
2GB-4GB memory, 256MB is reserved for crashkernel. In scenarios where
memory usage is high like nfs dump and usage of '-E' option in
makedumpfile, this triggers an OOM in crash kernel. To accomadate such
scenarios, crashkernel memory is increased from 256MB to 320MB.

Signed-off-by: Hari Bathini <hbathini@linux.vnet.ibm.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/machine_kexec.c b/arch/powerpc/kernel/machine_kexec.c
index 506bc24..1d82c50 100644
--- a/arch/powerpc/kernel/machine_kexec.c
+++ b/arch/powerpc/kernel/machine_kexec.c
@@ -129,10 +129,10 @@ unsigned long long __init arch_default_crash_size(unsigned long long total_size)
 #ifdef CONFIG_64BIT
  /*
   * crashkernel 'auto' reservation scheme
-  * 2G-4G:256M,4G-32G:512M,32G-64G:1G,64G-128G:2G,128G-:4G
+  * 2G-4G:320M,4G-32G:512M,32G-64G:1G,64G-128G:2G,128G-:4G
   */
  if (total_size < (1ULL<<32)) /* 4G */
-  return 1ULL<<28; /* 256M */
+  return ((1ULL<<28) + (1ULL<<26)); /* 320M */
  if (total_size < (1ULL<<35)) /* 32G */
   return 1ULL<<29; /* 512M */
  if (total_size < (1ULL<<36)) /* 64G */
-- 
1.7.1